clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -3091,6 +3104,58 @@ static void handleMaxGlobalWorkDimAttr(Sema &S, Decl *D,
                                                                      E);
 }
 
+SYCLIntelLoopFuseAttr *
+Sema::mergeSYCLIntelLoopFuseAttr(Decl *D, const SYCLIntelLoopFuseAttr &Attr,
+                                 Expr *E) {
+  if (checkAttrMutualExclusion<SYCLIntelLoopFuseIndependentAttr>(*this, D,
+                                                                 Attr))
+    return nullptr;
+
+  if (D->hasAttr<SYCLIntelLoopFuseAttr>())
+    return nullptr;
+
+  return ::new (Context) SYCLIntelLoopFuseAttr(Context, Attr, E);
+}
+
+SYCLIntelLoopFuseIndependentAttr *Sema::mergeSYCLIntelLoopFuseIndependentAttr(
+    Decl *D, const SYCLIntelLoopFuseIndependentAttr &Attr, Expr *E) {
+  if (checkAttrMutualExclusion<SYCLIntelLoopFuseAttr>(*this, D, Attr))
+    return nullptr;
+
+  if (D->hasAttr<SYCLIntelLoopFuseIndependentAttr>())
+    return nullptr;
+
+  return ::new (Context) SYCLIntelLoopFuseIndependentAttr(Context, Attr, E);
+}
+
+// Handles loop_fuse and loop_fuse_independent.
+// These attributes are incompatible with eachother.
+template <typename AttrType, typename ConflictingAttrType>
+static void handleLoopFusionAttr(Sema &S, Decl *D, const ParsedAttr &Attr) {
+  if (D->isInvalidDecl())
```

Comment:
I think we should actually drop this check -- attaching the attribute to an invalid declaration should be harmless, but failing to attach it may cause problems for the folks who are trying to keep erroneous AST nodes around for source fidelity reasons.